packages/react-ui-core/src/Gmap/FreeDrawLayer.js

```diff
@@ -0,0 +1,230 @@
+      shape.push(
+        poly.getPath()
+          .getAt(i)
+          .toUrlValue(5)
+          .split(',')
+          .map(Number)
+          .reverse()
+      )
+    }
+    return shape
+  }
+
+  enableMapControls() {
+    const { map, mapControls } = this.props
+    map.setOptions({
+      draggable: true,
+      zoomControl: true,
+      scrollwheel: true,
+      disableDoubleClickZoom: true,
+      ...mapControls,
+    })
+    this.setState({
+      drawing: false,
+    })
+  }
+
+  disableMapControls() {
+    const { map } = this.props
+    map.setOptions({
+      draggable: false,
+      zoomControl: false,
+      scrollwheel: false,
+      disableDoubleClickZoom: false,
+    })
+    this.setState({
+      drawing: true,
+    })
+  }
+
+  @autobind
+  enableMapDraw() {
+    const { map } = this.props
+    this.disableMapControls()
+    setupEvents(map, EVENTS, {
+      onMouseDown: () => {
+        this.drawFreeHand()
+      },
+    }, true)
+  }
+
+  drawFreeHand() {
+    const { map, onMapDrawStart, multipleShapes, dataStyle } = this.props
+    const poly = new window.google.maps.Polyline({
+      map,
+      clickable: false,
+      ...dataStyle.polyline,
+    })
+    const drawPoly = new window.google.maps.Polygon({
+      map,
+      path: poly.getPath(),
+      clickable: false,
+      ...dataStyle.polylineFill,
+    })
+
+    if (onMapDrawStart) {
+      onMapDrawStart()
+    }
+
+    if (!multipleShapes) {
+      this.clearAllShapes()
+    }
+
+    const listener = setupEvents(map, EVENTS, {
+      onMouseMove: e => {
+        poly.getPath().push(e.latLng)
+      },
+      onTouchStart: e => {
+        poly.getPath().push(e.latLng)
+      },
+    }, false)
+
+    setupEvents(map, EVENTS, {
+      onMouseUp: () => {
+        removeEvent(listener.onMouseMove)
+        poly.setMap(null)
+        drawPoly.setMap(null)
+        this.addShape(poly.getPath())
+        this.enableMapControls()
+      },
+      onTouchEnd: () => {
+        removeEvent(listener.onMouseMove)
+        poly.setMap(null)
+        drawPoly.setMap(null)
+        this.addShape(poly.getPath())
+        this.enableMapControls()
+      },
+    }, true)
+  }
+
+  addShapeToGMaps(path) {
+    const { map, dataStyle } = this.props
+    return new window.google.maps.Polygon({
+      map,
+      path,
+      ...dataStyle.polygon,
+    })
+  }
+
+  addShape(path) {
+    const { onMapDrawEnd } = this.props
+    const { polygons } = this.state
+
+    const rawPolygon = this.addShapeToGMaps(path)
+
+    this.setState({
+      polygons: [...polygons, rawPolygon],
+    }, () => {
+      if (onMapDrawEnd) {
+        onMapDrawEnd(this.state.polygons.reduce((acc, polygon, i) => {
+          acc[i] = this.getPolygonCoords(polygon)
+          return acc
+        }, {}))
+      }
+    })
+  }
+
+  @autobind
+  clearShape(id) {
+    const polygons = [...this.state.polygons]
+    polygons[id].setMap(null)
+    polygons.splice(id, 1)
+    this.setState({
```

Comment:
Maybe we should rename this as clearSingleShape. Although, I don't think we need this function since we might never even need this functionality. If product does decide to add this in the future we can add this function

Reply:
We need this function to call polygons[id].setMap(null)